### DISCUSSION TITLE
Adding custom attributes to Spree::Api::ApiHelpers in the initializer doesn't seem to work.

### DISCUSSION THREAD

**Discussion by rabbitbike** [solidusio/solidus#3956] [Troubleshooting] [answered]
I have an issue with adding custom attributes as return values in api get requests. I followed the edge documentation and added `Spree::Api::ApiHelpers.<custom>_attributes << [:attriubte1, :attribute2]` in initializers/spree.rb. The attributes get added, but once the client makes a get request from the browser and it reaches the server, I find only the default values in `Spree::Api::ApiHelpers.<custom>_attributes` in the `Api::BaseController`.

I can add strong parameters in the same way through `Spree::PermittedAttributes` in the initializer and it works so I can patch/post and persist custom attributes in db. I don’t know if it is related to how the modules are loaded and/or how the ApiHelpers is referenced in the `Api::BaseController` via `helper` method. 

I don’t know where else to look and would need help. I saw two other users in Slack forum raised the issue a while ago and hope anyone may have an answer. Thank you.

**Comment by kennyadsl** [comment, marked as answer]
Hey there :wave:, what is `custom` in `Spree::Api::ApiHelpers.<custom>_attributes`? I mean, is this for some custom model you created in your store or for something already present in Solidus core?

**Reply by rabbitbike** [reply]
I'm sorry I wasn't clear. For example, I added a `default_language` string field to the `Spree::User` model. So on the bottom of the initializers/spree.rb, I added the following line: `Spree::Api::ApiHelpers.user_attributes << :default_language` and restarted the server. The `default_language` field is still not being returned by the get request.

I would like to return its value via `GET https://example.com/api/users/{id}`, along with the default values which are `[:id, :email, :created_at, :updated_at]` and `bill_address` and `ship_address`.  

When I pause the code execution in `solidus_api-2.11/app/views/spree/api/users/_user.json.jbuilder`, `user_attributes` doesn't contain the `default_language`. Nor can I find it when I stop and inspect the `Spree::Api::BaseController`'s `helper Spree::Api::ApiHelpers` line.

I am in development environment and have not been able to test it test or production.

**Reply by kennyadsl** [reply]
What if you start a new Rails console and run `Spree::Api::ApiHelpers.user_attributes`, can you see it? If yes I think we can assume it's a code reloading issue.

**Reply by rabbitbike** [reply]
From the console, I don't see the attribute I wanted to add. 

```
$ rails console
Running via Spring preloader in process 33412
Loading development environment (Rails 6.0.3.4)
2.6.6 :001 > Spree::Api::ApiHelpers.user_attributes
 => [:id, :email, :created_at, :updated_at]
```

This is after many restarts. I tried adding a separate initializer file instead of initializers/spree.rb, but still the same result.

**Reply by rabbitbike** [reply]
I created a workaround by using a decorator. The following is the code:

```
module Decorators::ApiBaseControllerDecorator
  def self.prepended(base)
    base.before_action :add_custom_attributes
  end

  def add_custom_attributes
    unless Spree::Api::ApiHelpers.user_attributes.include? :default_language
      Spree::Api::ApiHelpers.user_attributes << :default_language
    end
  end

  Spree::Api::BaseController.prepend self
end
```

It is working but I still think it will be nice if we can just declare all custom fields for any models in the spree.rb initializer as suggested in the guide. Thank you.

**Reply by kennyadsl** [reply]
I just tried this and you are right, it is not working anymore for some reason. I think it's due to the new Rails autoloader somehow but couldn't find the root cause. If I'm right this doesn't happen in production and it's probably why it has not been reported yet.

I also started questioning why we have that method for setting parameters instead of just using a config for each model attributes we want to let people customize. I did an experiment and I think we can change the current code and start using a configuration for setting API response attributes. I'll leave some code here just as reference:

```ruby
# solidus/api/app/helpers/spree/api/api_helpers.rb

# Deprecate accessing the ATTRIBUTES constant

# for each attributes we can do 
def self.user_attributes
 Spree::Deprecation.warn("Please use Spree::Api::Config::user_attributes instead.")
 Spree::Api::Config.user_attributes
end
```

```ruby
# solidus/api/lib/spree/api_configuration.rb

module Spree
  class ApiConfiguration < Preferences::Configuration
    # ...

    preference :user_attributes, :array, default: [:id, :email, :created_at, :updated_at]
  end
end
```

```ruby
# host_app/config/initializers/spree.rb

Spree::Api::Config.configure do |config|
  config.user_attributes += [ :your_custom_attributes ]
end
```

And I confirm this works in development as well.

**Reply by rabbitbike** [reply]
Thank you! I will give it a try later!

**Reply by kennyadsl** [reply]
This is more something needed to change in the core, I jotted it down here just for reference. no need to try in your application. Thanks for reporting this issue!

**Reply by rabbitbike** [reply]
Thank you for the reference code. I tried it anyway, and I would like to report that I was able to make it work.

In my code I had to make a few arrangements. Since I would like to continue using the `solidus_api` gem as is, I was not able to directly modify the files in your reference code. Instead, I did the following:

```
# host_app/config/initializers/spree.rb

Spree::ApiConfiguration.preference :user_attributes, :array, default: [:id, :email, :created_at, :updated_at]

Spree::Api::Config.configure do |config|
  config.user_attributes += [ :your_custom_attributes ]
end

# host_app/app/helpers/decorators/api_helpers_decorator.rb

module Decorators::ApiHelpersDecorator
  Spree::Api::ApiHelpers.module_eval do
    def user_attributes
      Spree::Deprecation.warn("Please use Spree::Api::Config::user_attributes instead.")
      Spree::Api::Config.user_attributes
    end
  end
end
```

The reason I am assigning the preference in the spree.rb file is that it seems the spree.rb gets called before decorators. With these changes, I am able to get custom attributes through api calls. 

Thank you for solving the issue and I am super happy to hear that you are taking a note of this issue. Thank you!

**Reply by kennyadsl** [reply]
If you are monkey patching the `ApiHelpers` module in your application, I think you could even just do:

```ruby
Spree::Api::ApiHelpers.module_eval do
  user_attributes << :another_attribute
end
```

without all the rest of the changes I mentioned. Hope this helps to simplify your codebase.

**Reply by rabbitbike** [reply]
Yes, I think that is the way to go. It's simple and I can add fields for other models here as well, and it only gets called once. Thank you!

**Comment by kennyadsl** [comment]
I've opened this issue https://github.com/solidusio/solidus/issues/4027, hope someone can help us fix it.

**Reply by kennyadsl** [reply]
This has been done in #4039. 

Starting from Solidus 3.1, we can use the following approach:

```
Spree::Api::Config.configure do |config|
  config.user_attributes += [ :your_custom_attributes ]
end
```
